File: lib/core/services/push_notification_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:clashkingapp/core/app/my_app_state.dart';
import 'package:clashkingapp/core/config/api_config.dart';
import 'package:clashkingapp/core/config/app_feature_flags.dart';
import 'package:clashkingapp/core/constants/global_keys.dart';
import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/core/services/token_service.dart';
import 'package:clashkingapp/core/utils/debug_utils.dart';
import 'package:clashkingapp/common/widgets/dialogs/open_clash_dialog.dart';
import 'package:clashkingapp/firebase_options.dart';
import 'package:clashkingapp/features/pages/data/announcement_service.dart';
import 'package:clashkingapp/features/pages/data/announcement_story_cache_service.dart';
import 'package:clashkingapp/features/pages/presentation/announcement_story_dialog.dart';
import 'package:clashkingapp/features/pages/presentation/announcement_webview_page.dart';
import 'package:clashkingapp/features/pages/presentation/posts_page.dart';
import 'package:clashkingapp/features/pages/presentation/search_page.dart';
import 'package:clashkingapp/features/upgrade_tracker/presentation/upgrade_tracker_page.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:provider/provider.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:sentry_flutter/sentry_flutter.dart';

@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  try {
    await Firebase.initializeApp(
      options: DefaultFirebaseOptions.currentPlatform,
    );
    DebugUtils.debugInfo(
      'Push background message received: ${message.messageId}',
    );
  } catch (error, stackTrace) {
    await Sentry.captureException(error, stackTrace: stackTrace);
    DebugUtils.debugWarning('Push background handler skipped: $error');
  }
}

enum PushNotificationSetupState {
  unsupported,
  notConfigured,
  initializing,
  ready,
  permissionRequired,
  permissionDenied,
  tokenUnavailable,
}

class PushNotificationSetupResult {
  const PushNotificationSetupResult({
    required this.state,
    this.authorizationStatus,
    this.token,
    this.message,
  });

  final PushNotificationSetupState state;
  final AuthorizationStatus? authorizationStatus;
  final String? token;
  final String? message;

  bool get canReceivePush =>
      state == PushNotificationSetupState.ready && token != null;
}

class PushNotificationService {
  PushNotificationService._();

  static final PushNotificationService instance = PushNotificationService._();

  static const _pushApiV2BaseOverride = String.fromEnvironment(
    'CK_PUSH_API_V2_BASE_URL',
  );

  static String get _pushEnvironment {
    if (_pushApiV2BaseOverride.isNotEmpty ||
        ApiConfig.environment == ApiEnvironment.local) {
      return 'sandbox';
    }
    return 'production';
  }

  static void registerBackgroundHandler() {
    if (!kIsWeb && (Platform.isAndroid || Platform.isIOS)) {
      FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);
    }
  }

  static const _deviceEndpoint = '/notifications/devices';
  static const _preferencesEndpoint = '/notifications/preferences';
  static const _tokenPrefsKey = 'push_fcm_token';
  static const _lastRegistrationPrefsKey = 'push_last_registration_token';
  static const _channelId = 'clashking_push';
  static const _channelName = 'ClashKing alerts';
  static const _channelDescription =
      'War, CWL, account, and ClashKing announcement alerts.';

  final FlutterLocalNotificationsPlugin _localNotifications =
      FlutterLocalNotificationsPlugin();

  StreamSubscription<RemoteMessage>? _foregroundSub;
  StreamSubscription<RemoteMessage>? _openedSub;
  StreamSubscription<String>? _tokenRefreshSub;
  bool _initializing = false;
  bool _initialized = false;
  bool _localNotificationsReady = false;
  PushNotificationSetupResult _lastResult = const PushNotificationSetupResult(
    state: PushNotificationSetupState.notConfigured,
  );

  PushNotificationSetupResult get lastResult => _lastResult;

  Future<PushNotificationSetupResult> initialize({
    bool register = false,
  }) async {
    if (kIsWeb || !(Platform.isAndroid || Platform.isIOS)) {
      return _setResult(
        const PushNotificationSetupResult(
          state: PushNotificationSetupState.unsupported,
          message: null,
        ),
      );
    }

    if (_initializing) {
      return _lastResult;
    }

    _initializing = true;
    _setResult(
      const PushNotificationSetupResult(
        state: PushNotificationSetupState.initializing,
      ),
    );

    try {
      await _ensureFirebaseInitialized();
      await _initializeLocalNotifications();
      await FirebaseMessaging.instance
          .setForegroundNotificationPresentationOptions(
            alert: true,
            badge: true,
            sound: true,
          );

      _bindMessageStreams();
      _initialized = true;

      final settings = await FirebaseMessaging.instance
          .getNotificationSettings();
      if (!_hasDisplayPermission(settings.authorizationStatus)) {
        return _setResult(
          PushNotificationSetupResult(
            state: _permissionStateFor(settings.authorizationStatus),
            authorizationStatus: settings.authorizationStatus,
            message: null,
          ),
        );
      }

      final token = await FirebaseMessaging.instance.getToken();
      if (token == null || token.isEmpty) {
        return _setResult(
          const PushNotificationSetupResult(
            state: PushNotificationSetupState.tokenUnavailable,
            message: null,
          ),
        );
      }

      await _cacheToken(token);
      if (register) {
        unawaited(registerCurrentDeviceToken(token: token));
      }

      return _setResult(
        PushNotificationSetupResult(
          state: PushNotificationSetupState.ready,
          token: token,
        ),
      );
    } catch (error, stackTrace) {
      await Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugWarning('Push notifications are not configured: $error');
      return _setResult(
        PushNotificationSetupResult(
          state: PushNotificationSetupState.notConfigured,
          message: error.toString(),
        ),
      );
    } finally {
      _initializing = false;
    }
  }

  Future<PushNotificationSetupResult> requestPermissionAndRegister() async {
    final initialized = _initialized
        ? _lastResult
        : await initialize(register: false);
    if (initialized.state == PushNotificationSetupState.notConfigured ||
        initialized.state == PushNotificationSetupState.unsupported) {
      return initialized;
    }

    try {
      await FirebaseMessaging.instance.requestPermission(
        alert: true,
        announcement: false,
        badge: true,
        carPlay: false,
        criticalAlert: false,
        provisional: false,
        sound: true,
      );

      if (Platform.isAndroid) {
        await _localNotifications
            .resolvePlatformSpecificImplementation<
              AndroidFlutterLocalNotificationsPlugin
            >()
            ?.requestNotificationsPermission();
      }

      final resolvedSettings = await FirebaseMessaging.instance
          .getNotificationSettings();
      final authorizationStatus = resolvedSettings.authorizationStatus;

      if (!_hasDisplayPermission(authorizationStatus)) {
        return _setResult(
          PushNotificationSetupResult(
            state: _permissionStateFor(authorizationStatus),
            authorizationStatus: authorizationStatus,
            message: null,
          ),
        );
      }

      final token = await FirebaseMessaging.instance.getToken();
      if (token == null || token.isEmpty) {
        return _setResult(
          PushNotificationSetupResult(
            state: PushNotificationSetupState.tokenUnavailable,
            authorizationStatus: authorizationStatus,
            message: null,
          ),
        );
      }

      await _cacheToken(token);
      await registerCurrentDeviceToken(token: token);

      return _setResult(
        PushNotificationSetupResult(
          state: PushNotificationSetupState.ready,
          authorizationStatus: authorizationStatus,
          token: token,
        ),
      );
    } catch (error, stackTrace) {
      await Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugError('Push permission/register failed: $error');
      return _setResult(
        PushNotificationSetupResult(
          state: PushNotificationSetupState.notConfigured,
          message: error.toString(),
        ),
      );
    }
  }

  Future<void> registerCurrentDeviceToken({String? token}) async {
    final resolvedToken = token ?? await cachedToken();
    if (resolvedToken == null || resolvedToken.isEmpty) {
      DebugUtils.debugWarning('Push registration skipped: no FCM token.');
      return;
    }

    final packageInfo = await PackageInfo.fromPlatform();
    final tokenService = TokenService();
    final settings = await FirebaseMessaging.instance.getNotificationSettings();
    final payload = <String, dynamic>{
      'token': resolvedToken,
      'device_id': await tokenService.getDeviceId(),
      'provider': 'fcm',
      'platform': Platform.operatingSystem,
      'environment': _pushEnvironment,
      'app_version': packageInfo.version,
      'build_number': packageInfo.buildNumber,
      'os_version': Platform.operatingSystemVersion,
      'device_model': await tokenService.getDeviceName(),
      'authorization_status': settings.authorizationStatus.name,
      'locale': PlatformDispatcher.instance.locale.toLanguageTag(),
      'timezone': DateTime.now().timeZoneName,
    };

    if (Platform.isIOS) {
      final apnsToken = await FirebaseMessaging.instance.getAPNSToken();
      if (apnsToken != null) {
        payload['apns_token'] = apnsToken;
      }
    }

    try {
      final url = _pushApiV2BaseOverride.isEmpty
          ? null
          : '$_pushApiV2BaseOverride$_deviceEndpoint';
      final response = await ApiService().postResponse(
        _deviceEndpoint,
        body: payload,
        requiresAuth: true,
        url: url,
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_lastRegistrationPrefsKey, resolvedToken);
        DebugUtils.debugSuccess('Push device token registered.');
      } else {
        DebugUtils.debugWarning(
          'Push token registration failed: ${response.statusCode}',
        );
      }
    } catch (error, stackTrace) {
      // Backend endpoint may not exist yet; keep this non-fatal so local push
      // reception can be tested before server registration is wired.
      await Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugWarning('Push token registration skipped: $error');
    }
  }

  Future<bool> savePreferences(Map<String, dynamic> preferences) async {
    if (kIsWeb || !(Platform.isAndroid || Platform.isIOS)) return false;
    final payload = <String, dynamic>{
      'device_id': await TokenService().getDeviceId(),
      'environment': _pushEnvironment,
      'locale': PlatformDispatcher.instance.locale.toLanguageTag(),
      'timezone': DateTime.now().timeZoneName,
      'enabled_types': <String>[],
      'war_attack_modes': <String>[],
      'event_types': <String>[],
      'reminder_timings': <String>[],
      'account_scope': 'all',
      'selected_accounts': <String>[],
      'selected_town_halls': <int>[],
      'selected_clan_tags': <String>[],
      ...preferences,
    };
    try {
      final url = _pushApiV2BaseOverride.isEmpty
          ? null
          : '$_pushApiV2BaseOverride$_preferencesEndpoint';
      final response = await ApiService().putResponse(
        _preferencesEndpoint,
        body: payload,
        requiresAuth: true,
        url: url,
      );
      return response.statusCode >= 200 && response.statusCode < 300;
    } catch (error, stackTrace) {
      await Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugWarning('Push preferences sync failed: $error');
      return false;
    }
  }

  Future<bool> unregisterCurrentDeviceToken() async {
    if (kIsWeb || !(Platform.isAndroid || Platform.isIOS)) return false;

    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(_tokenPrefsKey);
    final payload = <String, dynamic>{
      'device_id': await TokenService().getDeviceId(),
      'provider': 'fcm',
      'platform': Platform.operatingSystem,
      'environment': _pushEnvironment,
      if (token != null && token.isNotEmpty) 'token': token,
    };

    try {
      final url = _pushApiV2BaseOverride.isEmpty
          ? null
          : '$_pushApiV2BaseOverride$_deviceEndpoint';
      final response = await ApiService().deleteResponse(
        _deviceEndpoint,
        body: payload,
        requiresAuth: true,
        url: url,
      );
      if (response.statusCode >= 200 && response.statusCode < 300) {
        await Future.wait([
          prefs.remove(_tokenPrefsKey),
          prefs.remove(_lastRegistrationPrefsKey),
        ]);
        await FirebaseMessaging.instance.deleteToken();
        return true;
      }
    } catch (error, stackTrace) {
      await Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugWarning('Push device unregister skipped: $error');
    }
    return false;
  }

  Future<String?> cachedToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenPrefsKey);
  }

  Future<String?> tokenPreview() async {
    final token = await cachedToken();
    if (token == null || token.length <= 16) return token;
    return '${token.substring(0, 8)}…${token.substring(token.length - 8)}';
  }

  Future<void> showForegroundMessage(RemoteMessage message) async {
    if (!_localNotificationsReady) return;

    final notification = message.notification;
    final title =
        notification?.title ?? message.data['title']?.toString() ?? 'ClashKing';
    final body =
        notification?.body ??
        message.data['body']?.toString() ??
        'New ClashKing alert';
    final payload = message.data.isEmpty ? null : jsonEncode(message.data);

    await _localNotifications.show(
      id: DateTime.now().millisecondsSinceEpoch ~/ 1000,
      title: title,
      body: body,
      notificationDetails: const NotificationDetails(
        android: AndroidNotificationDetails(
          _channelId,
          _channelName,
          channelDescription: _channelDescription,
          importance: Importance.high,
          priority: Priority.high,
          icon: '@drawable/ic_stat_clashking',
        ),
        iOS: DarwinNotificationDetails(
          presentAlert: true,
          presentBadge: true,
          presentSound: true,
        ),
      ),
      payload: payload,
    );
  }

  Future<void> _ensureFirebaseInitialized() async {
    if (Firebase.apps.isNotEmpty) return;
    await Firebase.initializeApp(
      options: DefaultFirebaseOptions.currentPlatform,
    );
  }

  Future<void> _initializeLocalNotifications() async {
    if (_localNotificationsReady) return;

    const initializationSettings = InitializationSettings(
      android: AndroidInitializationSettings('@drawable/ic_stat_clashking'),
      iOS: DarwinInitializationSettings(
        requestAlertPermission: false,
        requestBadgePermission: false,
        requestSoundPermission: false,
      ),
    );

    await _localNotifications.initialize(
      settings: initializationSettings,
      onDidReceiveNotificationResponse: (response) {
        _handlePayload(response.payload);
      },
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin
        >()
        ?.createNotificationChannel(
          const AndroidNotificationChannel(
            _channelId,
            _channelName,
            description: _channelDescription,
            importance: Importance.high,
          ),
        );

    _localNotificationsReady = true;
  }

  void _bindMessageStreams() {
    _foregroundSub ??= FirebaseMessaging.onMessage.listen((message) {
      DebugUtils.debugInfo('Push foreground message: ${message.messageId}');
      unawaited(showForegroundMessage(message));
    });

    _openedSub ??= FirebaseMessaging.onMessageOpenedApp.listen((message) {
      DebugUtils.debugInfo('Push opened message: ${message.messageId}');
      _handleMessageNavigation(message);
    });

    _tokenRefreshSub ??= FirebaseMessaging.instance.onTokenRefresh.listen((
      token,
    ) {
      DebugUtils.debugInfo('Push token refreshed.');
      unawaited(_cacheToken(token));
      unawaited(registerCurrentDeviceToken(token: token));
    });

    unawaited(
      FirebaseMessaging.instance.getInitialMessage().then((message) {
        if (message != null) {
          _handleMessageNavigation(message);
        }
      }),
    );
  }

  Future<void> _cacheToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenPrefsKey, token);
  }

  void _handlePayload(String? payload) {
    if (payload == null || payload.isEmpty) return;
    try {
      final decoded = jsonDecode(payload);
      if (decoded is Map<String, dynamic>) {
        _handleDataNavigation(decoded);
      }
    } catch (error, stackTrace) {
      Sentry.captureException(error, stackTrace: stackTrace);
      DebugUtils.debugWarning('Invalid notification payload: $error');
    }
  }

  void _handleMessageNavigation(RemoteMessage message) {
    _handleDataNavigation(message.data);
  }

  void _handleDataNavigation(Map<String, dynamic> data) {
    if (data['type']?.toString() == 'admin_post') {
      if (!_canOpenFeatureRoute('admin_post', AppFeatureFlags.posts)) return;
      final postID = data['post_id']?.toString();
      if (postID != null && postID.isNotEmpty) {
        unawaited(_openAdminPost(postID));
        return;
      }
    }
    final route = data['route']?.toString();
    if (route == null || route.isEmpty) return;
    unawaited(_openRoute(route));
  }

  Future<void> _openRoute(String route) async {
    var navigator = globalNavigatorKey.currentState;
    for (var attempt = 0; navigator == null && attempt < 20; attempt++) {
      await Future<void>.delayed(const Duration(milliseconds: 250));
      navigator = globalNavigatorKey.currentState;
    }
    if (navigator == null) return;

    switch (route) {
      case '/support-creator':
      case '/settings/support':
        final context = globalNavigatorKey.currentContext;
        if (context == null || !context.mounted) return;
        final languageCode = Localizations.localeOf(
          context,
        ).languageCode.toLowerCase();
        final url = Uri.https('link.clashofclans.com', '/$languageCode', {
          'action': 'SupportCreator',
          'id': 'Clashking',
        });
        await showDialog<void>(
          context: context,
          builder: (_) => OpenClashDialog(url: url),
        );
        return;
      case '/posts':
        if (!_canOpenFeatureRoute(route, AppFeatureFlags.posts)) return;
        await navigator.push(
          MaterialPageRoute<void>(builder: (_) => const PostsPage()),
        );
        return;
      case '/search':
        await navigator.push(
          MaterialPageRoute<void>(
            builder: (_) => const SearchPage(overlay: true, autofocus: true),
          ),
        );
        return;
      case '/upgrade-tracker':
        if (!_canOpenFeatureRoute(route, AppFeatureFlags.upgradeTracker)) {
          return;
        }
        await navigator.push(
          MaterialPageRoute<void>(builder: (_) => const UpgradeTrackerPage()),
        );
        return;
      default:
        DebugUtils.debugWarning('Unsupported push route: $route');
    }
  }

  bool _canOpenFeatureRoute(String route, String featureFlag) {
    if (_isFeatureEnabled(featureFlag)) return true;
    DebugUtils.debugWarning(
      'Push route blocked by feature flag: $route ($featureFlag)',
    );
    return false;
  }

  bool _isFeatureEnabled(String key) {
    final context = globalNavigatorKey.currentContext;
    if (context == null || !context.mounted) {
      return AppFeatureFlags.defaultValue(key);
    }
    try {
      return context.read<MyAppState>().isFeatureEnabled(key);
    } on ProviderNotFoundException {
      return AppFeatureFlags.defaultValue(key);
    }
  }

  Future<void> _openAdminPost(String postID) async {
    final announcement = await AnnouncementService().getAnnouncement(postID);
    if (announcement == null) {
      DebugUtils.debugWarning('Push post could not be loaded: $postID');
      return;
    }
    var navigator = globalNavigatorKey.currentState;
    for (var attempt = 0; navigator == null && attempt < 20; attempt++) {
      await Future<void>.delayed(const Duration(milliseconds: 250));
      navigator = globalNavigatorKey.currentState;
    }
    if (navigator == null) return;
    if (announcement.storyUrl?.isNotEmpty == true) {
      final preparedFilePath = await AnnouncementStoryCacheService().prepare(
        announcement,
      );
      if (preparedFilePath == null) {
        DebugUtils.debugWarning('Push story could not be prepared: $postID');
        return;
      }
      final storyContext = globalNavigatorKey.currentContext;
      if (storyContext == null || !storyContext.mounted) return;
      await showAnnouncementStoryDialog(
        storyContext,
        announcement: announcement,
        preparedFilePath: preparedFilePath,
      );
      return;
    }
    await navigator.push(
      MaterialPageRoute<void>(
        builder: (_) => AnnouncementWebViewPage(
          title: announcement.title,
          html: announcement.body,
          url: announcement.htmlUrl,
        ),
      ),
    );
  }

  PushNotificationSetupResult _setResult(PushNotificationSetupResult result) {
    _lastResult = result;
    return result;
  }

  bool _hasDisplayPermission(AuthorizationStatus status) {
    return status == AuthorizationStatus.authorized ||
        status == AuthorizationStatus.provisional;
  }

  PushNotificationSetupState _permissionStateFor(AuthorizationStatus status) {
    return status == AuthorizationStatus.denied
        ? PushNotificationSetupState.permissionDenied
        : PushNotificationSetupState.permissionRequired;
  }
}

File: expo/src/features/notifications/push/push-notification-service.ts
import type { ApiEnvironment, ApiRequestOptions } from '../../../core/api/client';
import { STORAGE_KEYS } from '../../../core/storage/storage';
import type {
  PushAuthorizationStatus,
  PushData,
  PushMessage,
  PushNotificationServiceOptions,
  PushNotificationSetupResult,
  PushPlatform,
  PushUnsubscribe,
  SupportedPushRoute,
} from './contracts';

export const PUSH_DEVICE_ENDPOINT = '/notifications/devices';
const ALL_HTTP_STATUSES = Array.from({ length: 500 }, (_, index) => index + 100);
const SUPPORTED_ROUTES = new Set<SupportedPushRoute>([
  '/support-creator',
  '/settings/support',
  '/posts',
  '/search',
  '/upgrade-tracker',
]);

export interface NotificationDeviceRegistrationPayload {
  readonly token: string;
  readonly device_id: string;
  readonly provider: 'fcm';
  readonly platform: 'ios' | 'android';
  readonly environment: 'sandbox' | 'production';
  readonly app_version: string;
  readonly locale: string;
  readonly authorization_status: NotificationDeviceAuthorizationStatus;
}

export type NotificationDeviceAuthorizationStatus =
  'authorized' | 'provisional' | 'denied' | 'not_determined';

export function apiAuthorizationStatus(
  status: PushAuthorizationStatus,
): NotificationDeviceAuthorizationStatus {
  if (status === 'notDetermined') return 'not_determined';
  // The current API has no App Clip-specific ephemeral value. An ephemeral
  // grant can display notifications, so register it as authorized.
  if (status === 'ephemeral') return 'authorized';
  return status;
}

export function pushEnvironment(
  environment: ApiEnvironment,
  pushApiV2BaseUrlOverride?: string,
): 'sandbox' | 'production' {
  return pushApiV2BaseUrlOverride?.trim()
    ? 'sandbox'
    : environment === 'production'
      ? 'production'
      : 'sandbox';
}

export function buildNotificationDeviceRegistrationPayload(input: {
  token: string;
  deviceId: string;
  platform: 'ios' | 'android';
  environment: 'sandbox' | 'production';
  appVersion: string;
  locale: string;
  authorizationStatus: PushAuthorizationStatus;
}): NotificationDeviceRegistrationPayload {
  return {
    token: input.token,
    device_id: input.deviceId,
    provider: 'fcm',
    platform: input.platform,
    environment: input.environment,
    app_version: input.appVersion,
    locale: input.locale,
    authorization_status: apiAuthorizationStatus(input.authorizationStatus),
  };
}

export function canReceivePush(result: PushNotificationSetupResult): boolean {
  return result.state === 'ready' && result.token !== undefined;
}

export function tokenPreview(token: string | null): string | null {
  if (token === null || token.length <= 16) return token;
  return `${token.slice(0, 8)}…${token.slice(-8)}`;
}

function hasDisplayPermission(status: PushAuthorizationStatus): boolean {
  return status === 'authorized' || status === 'provisional';
}

function permissionState(status: PushAuthorizationStatus) {
  return status === 'denied' ? 'permissionDenied' : 'permissionRequired';
}

function supportsPush(platform: PushPlatform): platform is 'ios' | 'android' {
  return platform === 'ios' || platform === 'android';
}

export class PushNotificationService {
  private initializing = false;
  private initialized = false;
  private localNotificationsReady = false;
  private subscriptions: PushUnsubscribe[] = [];
  private streamsBound = false;
  private lastSetupResult: PushNotificationSetupResult = {
    state: 'notConfigured',
  };

  constructor(private readonly options: PushNotificationServiceOptions) {}

  get lastResult(): PushNotificationSetupResult {
    return this.lastSetupResult;
  }

  get supportsPushNotifications(): boolean {
    return supportsPush(this.options.platform);
  }

  get environment(): 'sandbox' | 'production' {
    return pushEnvironment(this.options.apiEnvironment, this.options.pushApiV2BaseUrlOverride);
  }

  async initialize(register = false): Promise<PushNotificationSetupResult> {
    if (!this.supportsPushNotifications || this.options.runtime === undefined) {
      return this.setResult({
        state: this.supportsPushNotifications ? 'notConfigured' : 'unsupported',
      });
    }
    if (this.initializing) return this.lastSetupResult;

    this.initializing = true;
    this.setResult({ state: 'initializing' });
    try {
      if (!this.localNotificationsReady) {
        const unsubscribe = await this.options.runtime.initializeLocalNotifications((data) =>
          this.handleDataNavigation(data),
        );
        if (unsubscribe !== undefined) this.subscriptions.push(unsubscribe);
        this.localNotificationsReady = true;
      }
      this.bindMessageStreams();
      this.initialized = true;

      const authorizationStatus = await this.options.runtime.getAuthorizationStatus();
      if (!hasDisplayPermission(authorizationStatus)) {
        return this.setResult({
          state: permissionState(authorizationStatus),
          authorizationStatus,
        });
      }

      const token = await this.options.runtime.getToken();
      if (!token) {
        return this.setResult({ state: 'tokenUnavailable' });
      }

      await this.cacheToken(token);
      if (register) void this.registerCurrentDeviceToken({ token });
      return this.setResult({ state: 'ready', token });
    } catch (error) {
      await this.report('initialize', error);
      return this.setResult({ state: 'notConfigured', message: String(error) });
    } finally {
      this.initializing = false;
    }
  }

  async requestPermissionAndRegister(): Promise<PushNotificationSetupResult> {
    const initialized = this.initialized ? this.lastSetupResult : await this.initialize(false);
    if (initialized.state === 'notConfigured' || initialized.state === 'unsupported') {
      return initialized;
    }
    const runtime = this.options.runtime;
    if (runtime === undefined) return initialized;

    try {
      const authorizationStatus = await runtime.requestAuthorization();
      if (!hasDisplayPermission(authorizationStatus)) {
        return this.setResult({
          state: permissionState(authorizationStatus),
          authorizationStatus,
        });
      }
      const token = await runtime.getToken();
      if (!token) {
        return this.setResult({
          state: 'tokenUnavailable',
          authorizationStatus,
        });
      }
      await this.cacheToken(token);
      await this.registerCurrentDeviceToken({ token, allowDisabled: true });
      return this.setResult({ state: 'ready', authorizationStatus, token });
    } catch (error) {
      await this.report('permission', error);
      return this.setResult({ state: 'notConfigured', message: String(error) });
    }
  }

  async showPermissionPrimerOnce(onPermissionAccepted?: () => void | Promise<void>): Promise<void> {
    if (!this.supportsPushNotifications) return;
    const prompted = await this.options.preferences.getItem(
      STORAGE_KEYS.notificationPermissionPrimerShown,
    );
    if (prompted === 'true' || this.options.showPermissionPrimer === undefined) {
      return;
    }

    const shouldEnable = await this.options.showPermissionPrimer();
    await this.options.preferences.setItem(STORAGE_KEYS.notificationPermissionPrimerShown, 'true');
    if (!shouldEnable) return;

    const result = await this.requestPermissionAndRegister();
    if (canReceivePush(result) && onPermissionAccepted !== undefined) {
      try {
        await onPermissionAccepted();
      } catch (error) {
        await this.report('primer-callback', error);
      }
    }
  }

  async registerCurrentDeviceToken(
    options: {
      token?: string;
      allowDisabled?: boolean;
    } = {},
  ): Promise<void> {
    if (!supportsPush(this.options.platform) || this.options.runtime === undefined) {
      return;
    }
    if (!options.allowDisabled && !(await this.areNotificationsEnabled())) {
      this.options.log?.('Push registration skipped: notifications disabled.');
      return;
    }
    const token = options.token ?? (await this.cachedToken());
    if (!token) {
      this.options.log?.('Push registration skipped: no FCM token.');
      return;
    }

    try {
      const payload = buildNotificationDeviceRegistrationPayload({
        token,
        deviceId: await this.options.tokenService.getDeviceId(),
        platform: this.options.platform,
        environment: this.environment,
        appVersion: this.options.appVersion(),
        locale: this.options.locale(),
        authorizationStatus: await this.options.runtime.getAuthorizationStatus(),
      });
      const response = await this.request('POST', payload);
      if (response.status >= 200 && response.status < 300) {
        await this.options.preferences.setItem(STORAGE_KEYS.pushLastRegistrationToken, token);
        this.options.log?.('Push device token registered.');
      } else {
        this.options.log?.(`Push token registration failed: ${response.status}`);
      }
    } catch (error) {
      // Registration remains non-fatal while the API endpoint is unavailable.
      await this.report('register', error);
    }
  }

  async unregisterCurrentDeviceToken(): Promise<boolean> {
    if (!supportsPush(this.options.platform) || this.options.runtime === undefined) {
      return false;
    }

    let unregistered = false;
    try {
      const query = new URLSearchParams({
        device_id: await this.options.tokenService.getDeviceId(),
      });
      const response = await this.request(
        'DELETE',
        undefined,
        `${PUSH_DEVICE_ENDPOINT}?${query.toString()}`,
      );
      unregistered = response.status >= 200 && response.status < 300;
    } catch (error) {
      await this.report('unregister', error);
    } finally {
      await this.clearCurrentDeviceToken();
    }
    return unregistered;
  }

  cachedToken(): Promise<string | null> {
    return this.options.preferences.getItem(STORAGE_KEYS.pushFcmToken);
  }

  async areNotificationsEnabled(): Promise<boolean> {
    return (await this.options.preferences.getItem(STORAGE_KEYS.notificationsEnabled)) === 'true';
  }

  async tokenPreview(): Promise<string | null> {
    return tokenPreview(await this.cachedToken());
  }

  async showForegroundMessage(message: PushMessage): Promise<void> {
    if (!this.localNotificationsReady || this.options.runtime === undefined) return;
    await this.options.runtime.showLocalNotification(message);
  }

  dispose(): void {
    for (const unsubscribe of this.subscriptions.splice(0)) unsubscribe();
    this.streamsBound = false;
    this.localNotificationsReady = false;
    this.initialized = false;
  }

  private bindMessageStreams(): void {
    if (this.streamsBound || this.options.runtime === undefined) return;
    const runtime = this.options.runtime;
    this.streamsBound = true;
    this.subscriptions.push(
      runtime.onForegroundMessage((message) => {
        this.options.log?.(`Push foreground message: ${message.messageId ?? ''}`);
        void this.showForegroundMessage(message).catch((error) =>
          this.report('foreground-message', error),
        );
      }),
      runtime.onNotificationOpened((message) => this.handleDataNavigation(message.data ?? {})),
      runtime.onTokenRefresh((token) => {
        void this.handleTokenRefresh(token).catch((error) => this.report('register', error));
      }),
    );
    void runtime
      .getInitialMessage()
      .then((message) => {
        if (message !== null) this.handleDataNavigation(message.data ?? {});
      })
      .catch((error) => this.report('navigation', error));
    void runtime
      .getInitialLocalResponse()
      .then((data) => {
        if (data !== null) this.handleDataNavigation(data);
      })
      .catch((error) => this.report('navigation', error));
  }

  private async handleTokenRefresh(token: string): Promise<void> {
    if (!(await this.areNotificationsEnabled())) {
      this.options.log?.('Push token refresh ignored: notifications disabled.');
      return;
    }
    await this.cacheToken(token);
    await this.registerCurrentDeviceToken({ token });
  }

  private handleDataNavigation(data: PushData): void {
    if (String(data.type ?? '') === 'admin_post') {
      if (!this.options.isFeatureEnabled('posts')) return;
      const postId = String(data.post_id ?? '');
      if (postId) {
        void Promise.resolve(this.options.openAdminPost(postId)).catch((error) =>
          this.report('navigation', error),
        );
        return;
      }
    }

    const route = String(data.route ?? '');
    if (!SUPPORTED_ROUTES.has(route as SupportedPushRoute)) return;
    const supportedRoute = route as SupportedPushRoute;
    if (
      (supportedRoute === '/posts' && !this.options.isFeatureEnabled('posts')) ||
      (supportedRoute === '/upgrade-tracker' && !this.options.isFeatureEnabled('upgradeTracker'))
    ) {
      return;
    }
    void Promise.resolve(this.options.openRoute(supportedRoute)).catch((error) =>
      this.report('navigation', error),
    );
  }

  private async clearCurrentDeviceToken(): Promise<void> {
    const cleanupErrors: unknown[] = [];
    const removals = await Promise.allSettled([
      this.options.preferences.removeItem(STORAGE_KEYS.pushFcmToken),
      this.options.preferences.removeItem(STORAGE_KEYS.pushLastRegistrationToken),
    ]);
    for (const removal of removals) {
      if (removal.status === 'rejected') cleanupErrors.push(removal.reason);
    }
    this.setResult({ state: 'tokenUnavailable' });
    try {
      await this.options.runtime?.deleteToken();
    } catch (error) {
      cleanupErrors.push(error);
    }
    for (const error of cleanupErrors) await this.report('cleanup', error);
  }

  private cacheToken(token: string): Promise<void> {
    return this.options.preferences.setItem(STORAGE_KEYS.pushFcmToken, token);
  }

  private request(method: 'POST' | 'DELETE', body?: unknown, endpoint = PUSH_DEVICE_ENDPOINT) {
    const override = this.options.pushApiV2BaseUrlOverride?.replace(/\/$/, '');
    const requestOptions: ApiRequestOptions = {
      method,
      ...(body === undefined ? undefined : { body }),
      requiresAuth: true,
      acceptedStatuses: ALL_HTTP_STATUSES,
      ...(override ? { url: `${override}${endpoint}` } : undefined),
    };
    return this.options.api.request(endpoint, requestOptions);
  }

  private setResult(result: PushNotificationSetupResult) {
    this.lastSetupResult = result;
    return result;
  }

  private async report(
    operation: Parameters<NonNullable<typeof this.options.reportError>>[0]['operation'],
    error: unknown,
  ): Promise<void> {
    try {
      await this.options.reportError?.({ operation, error });
    } catch {
      // Reporting must never change notification or logout behavior.
    }
  }
}
